### PR TITLE
feat: indices are now versioned

### PR DESCRIPTION
**Pull request by simlu**

### CODE REVIEW

File: src/util/rest/data/count.js [hunk 1]
```diff
@@ -1,2 +1,3 @@
-module.exports = (call, idx) => call("POST", idx, { endpoint: "_count" })
-  .then(r => (r.statusCode === 200 ? r.body.count : false));
+module.exports = (call, idx) => call("POST", `${idx}@*`, { endpoint: "_count" })
+  // eslint-disable-next-line no-underscore-dangle
+  .then(r => (r.statusCode === 200 && r.body._shards.total !== 0 ? r.body.count : false));
```
Comment (MacMcIrish) [inline src/util/rest/data/count.js]: Does it make sense to use a `get(r, 'body._shards.total')`? You'd escape the `no-underscore-dangle` and avoid a 502 if `_shards` is not returned

File: test/index.spec.js [hunk 2]
```diff
@@ -125,16 +125,90 @@ describe('Testing index', () => {
...
+
+    // initialization
+    await index.rest.call("DELETE", `${indexName}@*`);
+    const create1 = await index.rest.call('PUT', `${indexName}@1`, { body: meta });
+    expect(create1.statusCode).to.equal(200);
+    const create2 = await index.rest.call('PUT', `${indexName}@2`, { body: meta });
+    expect(create2.statusCode).to.equal(200);
+
+    // add data
+    const id1 = uuid4();
```
Comment (MacMcIrish) [inline test/index.spec.js]: Why `id1`?

File: test/index.spec.js [hunk 3]
```diff
@@ -125,16 +125,90 @@ describe('Testing index', () => {
...
+      // create new index
+      expect(await index.rest.mapping.create("offer")).to.equal(true);
+      await validate(0, {});
+      // insert data
+      expect(await index.rest.data.update("offer", { upsert: uuids.map(id => ({ id })) })).to.equal(true);
+      await validate(3, {});
+      // create new version of index
+      index.index.register("offer", Object.assign({}, indices.offer, { fields: ["id"] }));
+      expect(await index.rest.mapping.create("offer")).to.equal(true);
+      await validate(3, { "offer@409492a32436d2cec4a3d01046308ae7e53893f1": 3 });
```
Comment (MacMcIrish) [inline test/index.spec.js]: This hash is going to change whenever we update the offer index, right? Would it make sense to calculate this in the setup? `currentOfferHash`

File: src/util/rest/data/update.js [hunk 4]
```diff
@@ -1,17 +1,34 @@
-module.exports = (call, idx, { remove = [], upsert = [] }) => {
+const historic = require("./../mapping/historic");
+
+module.exports = async (call, idx, mapping, { remove = [], upsert = [] }) => {
+  const oldVersionsEntries = Object.entries(await historic(call, idx, mapping));
+  const oldVersionsEmpty = oldVersionsEntries.filter(([_, v]) => v === 0).map(([k, _]) => k);
+  const oldVersionsNonEmpty = oldVersionsEntries.filter(([_, v]) => v !== 0).map(([k, _]) => k);
```
Comment (MacMcIrish) [inline src/util/rest/data/update.js]: Change v -> count or something more descriptive

File: src/util/rest/data/update.js [hunk 5]
```diff
@@ -1,17 +1,34 @@
...
+
+  // delete old, empty versions
+  await Promise.all(oldVersionsEmpty.map(i => call('DELETE', i)));
```
Comment (MacMcIrish) [inline src/util/rest/data/update.js]: Wrap in `if` statement

File: test/index.spec.js (hunk 2 above)
Comment (simlu) [reply test/index.spec.js]: fixed

File: test/index.spec.js (hunk 3 above)
Comment (simlu) [reply test/index.spec.js]: It will. I'm too lazy to do that now. Let's see how often this breaks :)

Comment (simlu) [reply test/index.spec.js]: I lied. Fixing now

**Pull request merged**